### PR TITLE
MIDI-97: delete unnecessary tokens

### PR DESCRIPTION
**Pull request by WojciechMat**
I have deleted padding token from project logic. 
### bos and eos tokens
Removing \<s> and \</s> special tokens turned out to be more complicated than I anticipated. They play a valuable role in attention mechanism. The attention layer finds correlation between current and previous tokens, so existance of start token seem like necessity for now:

image

When using greedy_decode we do need to pass starting token to the model at the beggining, so that attention mechanism has any sentence (composed of one token during first step) to work with. 

We could think of a way to overcome this and remove \<s> and \</s> tokens completly but I would have to think about it more as I do not see a viable solution right now. I will try to at least get rid of \</s> token now.

### CODE REVIEW

**Comment by roszcz** [comment]
I'm not convinced 😅 the chart you posted is from a language example, which does not apply in our case, because all of our "sentences" have exactly the same length (`sequence_len`), and we don't need additional tokens signal a change between sentences.

For the greedy decode, it should be possible to start with an empty sequence (attention_mask all negative?). If not, we just need to come up with some strategy for an initial velocity/dstart 🤔

**Comment by WojciechMat** [comment]
Starting with an empty sequence is not possibe, because we want to pass the starting sentence to https://github.com/Nospoko/midi-translation/blob/d80b24a43b6f76179d2cbe248aa7cabe34937e05/modules/decoder.py#L35
Which will return an empty tensor.
Unless of course I can think of a bypass.
I am not convinced as to starting velocity, because it will not depend on  the first src token at all - it does now. Unless we make it so :thinking:

**Comment by WojciechMat** [comment]
Deleted \</s> tokens. Thinking what to do with \<s>.

**Comment by roszcz** [comment]
I see, let's rename it to "<CLS>" for now - this will be consistent with BERT approach. We can later try to use it to do composer classification or other downstream tasks

**Comment by WojciechMat** [comment]
I moved src_mask generation in evaluations/dashboard to utils/greedy_decode and utils/decode_and_out, as they are always the same.
https://github.com/Nospoko/midi-translation/blob/338b05596740575d0c466ad2d2450af38749812c/utils.py#L114-L115

**Comment by WojciechMat** [comment]
I moved src_mask creation to utils:
https://github.com/Nospoko/midi-translation/blob/bd9428c8579b704e99b5b94ca6b3c34ea1529896/utils.py#L140

and updated src_mask creation in Batch class, taking advantage of the fact we do not use any padding.

File: data/tokenizer.py [hunk 1]
```diff
@@ -184,14 +174,10 @@ def quantize(self, start: list[float]) -> list[int]:
...
     def tokenize(self, record: dict) -> list[str]:
-        # TODO I don't love the idea of adding tokens durint *tokenize* call
-        # If we want to pretend that our midi sequences have start and finish
-        # we should take care of that before we get here :alarm:
         dstart_bins = self.quantize(record["start"])
 
         # get tokens from quantized data
-        tokens = [str(dstart_bin) for dstart_bin in dstart_bins]
-        tokens = ["<s>"] + tokens + ["</s>"]
+        tokens = [""] + [str(dstart_bin) for dstart_bin in dstart_bins]
```
Comment (roszcz) [inline data/tokenizer.py]:
```suggestion
        tokens = [str(dstart_bin) for dstart_bin in dstart_bins]
```

Removing my comment about not adding tokens in this function does not address the issue 😅 

I think special tokens should be added in a step separated from tokenization of `dstart/velocity` bins, let me know if you see good reasons not to do this, because maybe I'm missing something.

Also, this first token shouldn't be an empty string, but rather a `"<CLS>" `token. This will make it consistent with other transformer-oriented research that we're doing.

**Comment by WojciechMat** [comment]
How about I also delete doc.py and documentation folder as they were just a prototype for blog?

I changed the weird and obscure way of counting tokens for a simple tgt_y.numel().

File: data/tokenizer.py (hunk 1 above)
Comment (WojciechMat) [reply data/tokenizer.py]: Sorry, I deleted it when I was removing all the special tokens and thought it is resolved :sweat_smile: 

As for first token - I might have misunderstood your previeous comment - github hid the cls token :0.
> I see, let's rename it to "" for now - this will be consistent with BERT approach. We can later try to use it to do composer classification or other downstream tasks

**Comment by WojciechMat** [comment]
cls tokens are now added inside \_\_getitem__ method of the dataset.

**Review by roszcz — APPROVED** [review]
🚀 

This is not related to changes in this PR, but I don't like the name `.untokenize` we use in encoders - let me know if you can come up with a better name for this 🤔

**Comment by WojciechMat** [comment]
Perhabs we do not need this method at all - mayve we can replace it with MidiEncoder.decode and have it „untokenizing” directly from token ids

**Comment by roszcz** [comment]
My priority here is to keep it as easy to read as possible - in terms of the data flow, I think we have a very clear transition `midi->pandas->quantized->tokens->token_ids` (and back). It would be great to always know which step and which direction is any procedure responsible for. 

With "untokenize" I can only guess it's the second half, going "back". "Decode" for me would be something like "tokens -> pandas". 🤔

**Comment by WojciechMat** [comment]
What untokenize does is `tokens -> pandas` in QuantizedMidiEncoder or `tokens -> list[int]` in velocity and dstart encoders

We could rename it to decode or make decode take a list of `token_ids` and produce the same outcome.

There already is an implementation of decode in MidiEncoder, we just do not use it I think - it calls .untokenize after changing ids to tokens.

**Comment by roszcz** [comment]
Let's merge it as is, recognizing and naming all the entities here is important, but it can wait 👍
